File: src/previous/day22/BFSGraphMatrix.ts
// 8

export default function bfs(
    graph: WeightedAdjacencyMatrix,
    source: number,
    needle: number,
): number[] | null {
    const seen: boolean[] = new Array(graph.length).fill(false)
    const prev: number[] = new Array(graph.length).fill(-1)
    const queue: number[] = [source]

    seen[source] = true

    while (queue.length) {
        const curr = queue.shift() as number
        if (curr === needle) break

        const adjs = graph[curr]
        for (let i = 0; i < adjs.length; i++) {
            if (adjs[i] === 0 || seen[i]) continue
            prev[i] = curr
            seen[i] = true
            queue.push(i)
        }
    }
    const out: number[] = []
    let curr = needle
    if (prev[curr] === -1) return null

    while (prev[curr] !== -1) {
        out.push(curr)
        curr = prev[curr]
    }

    out.push(source)
    return out.reverse()
}

File: go/day01/BFSGraphMatrix/BFSGraphMatrix.go
package bfs_graph_matrix

import (
	data_structs "github.com/pakkermandev/kata-machine-learning/DataStructs"
)

func BFSGraphMatrix(graph data_structs.WeightedAdjacencyMatrix, source, needle int) []int {
	queue := []int{source}
	prev := make([]int, len(graph))
	seen := make([]bool, len(graph))

	for i := 0; i < len(graph); i++ {
		prev[i] = -1
		seen[i] = false
	}

	seen[source] = true
	for len(queue) != 0 {
		curr := queue[0]
		queue = queue[1:]

		adjs := graph[curr]
		for i := range adjs {
			if adjs[i] == 0 || seen[i] {
				continue
			}

			seen[i] = true
			prev[i] = curr
			queue = append(queue, i)

		}

	}

	var out []int
	curr := needle
	for prev[curr] != -1 {
		out = append(out, curr)
		curr = prev[curr]
	}

	out = append(out, source)

	var reverse []int
	for i := 0; i < len(out); i++ {
		reverse = append(reverse, out[len(out)-i-1])
	}

	return reverse
}
